### ISSUE TITLE
ORA-07455 when comparing sysrefcursor using exclude.

### ISSUE THREAD

**Issue by hinjo89** [utPLSQL/utPLSQL#752]
Hi,

I'm trying to introduce utPLSQL as Unit Testing framework in our PL/SQL components, since we currently don't have real unit testing.

Unfortunately I have a pretty big issue that's blocking me currently:

I'm trying to compare two cursors that only contain 2 rows (or even only one) and exclude one or two fields in the comparison.

Unfortunately, most of the time my database doesn't execute the comparison query in the package ut_compound_data_helper with the error ORA-07455: estimated execution time (2472326 secs), exceeds limit (115200 secs).

Of course it actually only takes miliseconds when it, out of whatever reason, isn't blocked by this.

I can't find out what's the reason for this, I suppose it's the row_number() function in combination with the XMLTYPE? Did you already experience this and have some hints on how to resolve it?

I'm not sure if an Issue created here is the right place for this, but I didn't know where else to ask this. If you could point me to a better place that would also be appreciated!

Thanks a lot,
Johannes

**Comment by jgebal** [comment]
What version of oracle are you on?
What patches are applied?
Is it Exadata, RAC or standalone DB?

**Comment by jgebal** [comment]
Also what versipn of UTPLSQL 3 are you using?

**Comment by jgebal** [comment]
What version of utPLSQL v3 are you on?

**Comment by mathewbutler** [comment]
Hi - can you check if you have a resource plan directive set, and if so what is the value of max_est_exec_time?

If this is a resource plan directive constraint then you as a workaround you may be able to creat a new plan directive that is unconstrained by execution time and assign users that do unit tests by to a consumer group that uses this plan. This isn’t my area of expertise so is worth you speaking to your DBA.

Another avenue to investigate is whether the source tables you are using have up to date statistics. 

A final avenue may be to add hints to  the cursors being compared such that the real cardinalitybis understood. You could try a dynamic sampling at level 2, or a cardinality hint. 

If the hints work then there may be an opportunity to build something into utPLSQL as an optional setting for this kind of environment. 

Let’s understand the cause and take things from there. 

Cheers.

**Comment by hinjo89** [comment]
Hi,
thanks a lot for the fast replies! For the environment information (sorry, should've included that first), they are as follows:

DB version: 12.1.0.2.0
Patches: I'm not sure how to find out all, but in dba_registry_sqlpatch I can see that we have patch 160419.
Our Dev, which is what I use for testing, is a standalone DB (non Exadata).
utPLSQL version: 3.1.2

Regarding the resource plan directive: Yes, this is set and the max_est_exec_time is set to the limit specified in my error message. I fear that currently I can't make my DBAs remove that limit for my user, but that's also something that I will try.
I actually updated the statistics of the source tables that I used, but what I discovered afterwards is, that in the query that has the problem (which is in ut_compound_data_helper.get_columns_diff), the actual table is not even used, but we are working solely with XML-tables that were already prepared in a previous function. So statistics, and with that hints, don't seem to be something that could be of use there.
Interesting: I saw that there is a new version of ut_compound_data_helper in one of the feature branches that introduces aliases in get_columns_filter for the xpath that is null and I added the new version of that function into my local utplsql version. After I did that it started to work for a while, but after around one day it was back to error 07455.

Thanks,
Johannes

**Comment by hinjo89** [comment]
Hi all,

I think I was able to solve my problem by adding a cardinality hint in UT_COMPOUND_DATA_HELPER. Thanks for the tip regarding the hint @mathewbutler, even though I didn't use it in the cursor itself.

The code in the function get_columns_diff at the start now looks like this in my UT-installation (the hint being the only thing added):

`l_sql := q'[
      with
        expected_cols as ( select /*+ cardinality(item_data 10) */ :a_expected as item_data from dual ),
        actual_cols as ( select /*+ cardinality(item_data 10) */ :a_actual as item_data from dual ),
  expected_cols_info as (`

Of course this is not something that could be done statically just like this in the real application, but maybe more dynamically by inserting a count of the XML-table into the place of the sql that is now a 10 in my example?

I will watch the behaviour for a while and report if this really solved the problem reliably.

Thanks a lot for your inputs and if you have still more ideas I'd be very interested!

Regards,
Johannes

**Comment by hinjo89** [comment]
Update: The Hint was not yet in the right spot in the sql, which caused the error to happen again after some time, but I think I found the correct spot now to reliably put the cardinality into the right spot. Will post it after it ran successfully for a couple of days.

**Comment by jgebal** [comment]
I would put it here:
```
SELECT /*+ CARDINALITY(xt 100) */ rownum expected_pos,
                   xt.name expected_name,
                   xt.type expected_type
....
SELECT /*+ CARDINALITY(xt 100) */ rownum actual_pos,
                   xt.name actual_name,
                   xt.type actual_type
```
Let us know if that solution works for you correctly and is proving stable across time.

**Comment by hinjo89** [comment]
Yes, that's where I ended up putting it yesterday. Will let you know after some days of working with it how it is over time. Thanks a lot!

**Comment by hinjo89** [comment]
Hi,

the solution with the CARDINALITY-Hint works. In my local version I enhanced it a bit by adding two optional parameters for the count of the two cursors to get_column_diff which you get from elements_count in the  ut_data_value(_refcursor) object. Those parameters are then used instead of the hard coded '100' in the above example. For refcursors they are handed over to get_column_diff in ut_data_value_refcursor.

If that makes sense I would be interested in contributing, but it might take some time with me figuring out how to set everything up :)

Thanks,
Johannes

**Comment by lwasylow** [comment]
I would be careful with using hints there. Its vere dependent on local env and in others can lead to choosing suboptimal plans. However there are few things we can look to improve underlying query without hints I believe or at least we should try.

**Comment by jgebal** [comment]
I agree with @hinjo89 on this one @lwasylow.
Adding `cardinality` hint in a query that does nothing more than a full outer join on to PLSQL tables cant do any harm, but will bring better estimates.
Oracle doesn't have much choice when choosing an execution plan anyway, as the in-memory PLSQL tables are not indexed, so either HASH JOIN or a SORT-MERGE JOIN would be used. 
If Oracle would choose to use NESTED-LOOPS, that could cause significant performance degradation with the growth of collection sizes.

I did some experiments with different join approaches for PLSQL tables some time ago, taking a look at cost of nested-loops on non-indexed PLSQL tables (https://github.com/jgebal/collection_joins)

**Comment by hinjo89** [comment]
Hi @jgebal ,

I checked your commit and I think that the cardinality hint should actually be at the spot that you suggested in your previous answer directly in the select from the xmltable. I tried the other spots, they didn't affect the overall estimate.
